### PR TITLE
test(nayduck) - Added a test for chain recovery after a really slow chunk

### PR DESCRIPTION
**Pull request by wacban**
Added a test where one chunk takes a really long time to apply. This is to simulate extreme undercharging. Under those circumstances we want the blockchain to be able to recover - to eventually start processing chunks in the affected shard again. In the worst case scenario the 

This test currently fails on nightly where stateless validation is enabled. This needs to be properly investigated and resolved. My suspicion is that the chain cannot get a new chunk in after the slow chunk. I think every time a chunk producer attempts to produce a new chunk, it* takes too long and it doesn't make it to the next block. Once the next block is produced the next chunk producer needs to start from scratch. 

  `*` "it"  can actually be different things. It could be that just producing the chunk after the slow one is too slow. Or it may actually be produced but then it's too slow to validate it - the validators set changes every block and needs to apply the slow chunk. Or it could be something else too on the end to end path of producing chunks and including chunks into blocks. 
  
@akhi3030 and @nagisa Can you have a look at the sleep host function and advise how to properly gate it? 

@tayfunelmas and @robin-near Can you have a look at the python test? 

For reference here is a sample output https://gist.github.com/wacban/7db697121b7fbeb64c392224a05dbbda

### CODE REVIEW

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 1]
```diff
@@ -1223,6 +1223,15 @@ impl<'a> VMLogic<'a> {
         self.gas_opcodes(gas)
     }
 
+    // TODO!!! Hide this behind a feature flag!
+    pub fn sleep_nanos(&mut self, nanos: u64) -> Result<()> {
```
Comment (akhi3030) [inline runtime/near-vm-runner/src/logic/logic.rs:1227]: I wonder if we can just use the compile time feature gating for this?  Or would that not work for your usecase?

Comment (wacban) [reply runtime/near-vm-runner/src/logic/logic.rs:1227]: I think it would here, but doesn't in the imports.rs because it's inside of some macro.

Comment (akhi3030) [reply runtime/near-vm-runner/src/logic/logic.rs:1227]: https://github.com/near/nearcore/pull/9317/files#diff-e9b653ec008c4aeece3871e171113ef743d50a1d8229bddf14d1ea4c51391675R286 is an example of how to do it in `imports.rs`?

Comment (wacban) [reply runtime/near-vm-runner/src/logic/logic.rs:1227]: Ah nice, I missed that, and there is another example with "sandbox" just above. I'll try this, thanks!

File: runtime/near-vm-runner/src/imports.rs [hunk 2]
```diff
@@ -284,6 +284,13 @@ imports! {
     // #  Sandbox  #
     // #############
     ##["sandbox"] sandbox_debug_log<[len: u64, ptr: u64] -> []>,
+
+    // Sleep for the given number of nanoseconds. This is the ultimate
+    // undercharing function as it doesn't consume much gas but takes a lot of
```
Comment (tayfunelmas) [inline runtime/near-vm-runner/src/imports.rs:289]: nit. typo

File: runtime/near-vm-runner/src/imports.rs [hunk 3]
```diff
@@ -284,6 +284,13 @@ imports! {
...
+    // time to execute. It must always be gated behind a feature flag and it
+    // must never be released to production.
+    // TODO!!! Hide this behind a feature flag!
```
Comment (tayfunelmas) [inline runtime/near-vm-runner/src/imports.rs:292]: nit. I guess adding this TODO to a single place would suffice, maybe here.

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 1 above)
Comment (nagisa) [reply runtime/near-vm-runner/src/logic/logic.rs:1227]: I think its better to always use parameters for contract runtime nowadays. You can follow the example of yield_resume from https://github.com/near/nearcore/commit/47020d42f44ab37eeca65483a4d61b0e7b9f673d.

File: runtime/near-vm-runner/src/logic/logic.rs [hunk 4]
```diff
@@ -1223,6 +1223,15 @@ impl<'a> VMLogic<'a> {
...
+        let duration = std::time::Duration::from_nanos(nanos);
+        eprintln!("boom sleeping for {duration:?}!");
```
Comment (nagisa) [inline runtime/near-vm-runner/src/logic/logic.rs:1229]: The host functions are already traced, so these should probably go, or at least be replaced with `tracing::warn` or somesuch.

**Review by nagisa — COMMENTED** [review]
One thing for `sleep_nanos` is that I think we should introduce a compute cost for it too. I don't quite know off hand how to do it however, but the logging function does have a compute cost associated with it, so following that example should be straightforward enough.

**Comment by wacban** [comment]
> One thing for `sleep_nanos` is that I think we should introduce a compute cost for it too. I don't quite know off hand how to do it however, but the logging function does have a compute cost associated with it, so following that example should be straightforward enough.

@nagisa Why do you think we should add compute costs? My idea for this method was to be as undercharged as possible and for that I want as little gas and compute as possible.

File: pytest/tests/sanity/slow_chunk.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+    # Spin up multiple validator nodes in a multi shard chain. Deploy a contract
+    # to one of the shards and call a function that sleeps for a long time.
+    # Check that the shard is able to recover and that new chunks appear.
+    def test(self):
+        # The number of validators and the number of shards.
+        n = 4
+
+        # The validators are single shard tracking, the rpc tracks all shards.
+        client_config_changes = {n: {"tracked_shards": [0]}}
```
[the remark this answers is not in the document]
Comment (wacban) [reply pytest/tests/sanity/slow_chunk.py:31]: good point, it's non-obvious, let me make it cleaner
Nodes in range [0, n) are validators, node n is rpc. So here I keep the default setting of empty tracked shards for the validators (single shard tracking) and only make the rpc node track all shards. I'll make that explicit and add a better comment.

File: runtime/near-vm-runner/src/imports.rs (hunk 3 above)
Comment (wacban) [reply runtime/near-vm-runner/src/imports.rs:292]: Those need to be addressed before merging, I'll do that shortly.

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 4 above)
Comment (wacban) [reply runtime/near-vm-runner/src/logic/logic.rs:1229]: Ah I didn't mean to merge that, it's just my debug logs. I'll remove that.

**Comment by nagisa** [comment]
> Why do you think we should add compute costs? My idea for this method was to be as undercharged as possible and for that I want as little gas and compute as possible.

To make it more flexible, multi-purpose in a way and to allow multiple layers of safeguards against accidental leak of the method (the baseline parameters would set 300TGas + minutes of compute cost for the method, so even if it did get enabled by accident, it would still fail to execute).

But I guess if you're looking to use the pytest framework, you don't actually have a way to specify what VMConfig looks like... so whatever lets keep it as-is.

**Review by akhi3030 — APPROVED** [review]
LGTM.

File: runtime/near-vm-runner/src/logic/logic.rs (hunk 1 above)
Comment (nagisa) [reply runtime/near-vm-runner/src/logic/logic.rs:1227]: Not really. The parameter files are baked into the binary at compile time. From within rust code this is pretty easy, all these parameters end up as fields in `VMConfig` that gets passed around all over. From outside I don't think there's a way unless we modify the interface of the neard binary in some way.

I don't think adding another compile time feature is the right way to go about solving this problem, but I also don't know how to solve it in a straightforward way that allows to achieve what you want in pytests.

**Review by tayfunelmas — APPROVED** [review]
LGTM, good to have this test in place to investigate the issue.

**Pull request merged**